### PR TITLE
done tutorial

### PR DESCRIPTION
**Pull request by Nicolas-Bertaud**

### CODE REVIEW

File: .idea/.gitignore
Comment (eschalkargans) [inline .idea/.gitignore:1]: Maybe you can exclude the `.idea` folder by adding it to the `.gitignore` file at root

File: resources/polygons.geojson [hunk 1]
```diff
@@ -3,8 +3,10 @@
...
-{ "type": "Feature", "properties": { "id": null, "class": "FARMLAND", "color": "Green" }, "geometry": { "type": "MultiPolygon", "coordinates": [ [ [ [ 0.804435668608206, 43.774709103468517 ], [ 0.844193029411034, 43.775466280235008 ], [ 0.865770743441819, 43.739989613092405 ], [ 0.831338832243599, 43.721246898232536 ], [ 0.76424372591599, 43.728847383084364 ], [ 0.765078925893384, 43.751864880600344 ], [ 0.804435668608206, 43.774709103468517 ] ] ] ] } }
+{ "type": "Feature", "properties": { "id": null, "class": "WATER", "color": "blue" }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 1.428090901495726, 43.602284563088126 ], [ 1.428089026885723, 43.602356545315459 ], [ 1.429040215614357, 43.603881791155978 ], [ 1.434488292290918, 43.60395654390355 ], [ 1.440126601590054, 43.600505239754945 ], [ 1.439821106746437, 43.593156244512485 ], [ 1.436059506933373, 43.593032827932589 ], [ 1.438266986990658, 43.59961574857978 ], [ 1.434837726552646, 43.601945105233732 ], [ 1.428090901495726, 43.602284563088126 ] ] ] } },
```
Comment (eschalkargans) [inline resources/polygons.geojson:6]: I see that you changed an existing `polygons.geojson` ; would it be possible to revert this change, and give a different name to this file, to preserve the original one?
You can also create a subfolder in `resources/nbertaud` for instance to store all your files

File: src/training_raster_clipper/implementation/nbertaud.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        xr.DataArray: A DataArray containing the 3 RGB bands from the visible spectrum
+    """
+
+    # Find the path of the folder with the correct resolution
+    spatial_resolution_location = list(sentinel_product_location.glob(f"GRANULE/*/IMG_DATA/R{resolution}m"))[0]
+
+    # Find the path of the bands
+    band_locations = {}
+    for band_name in band_names:
+        band_locations[band_name] = list(spatial_resolution_location.glob(f"*_{band_name}_*"))[0]
```
Comment (eschalkargans) [inline src/training_raster_clipper/implementation/nbertaud.py:51]: you can do it in one expression with  a dict comprehension eg

instead of A (imperative)

```python
band_locations = {}
for band_name in band_names:
    d[band_name] = f(band_name)
```

you can do B (declarative)

```python
band_locations = { band_name: f(band_name) for band_name in band_names }
```

File: src/training_raster_clipper/implementation/nbertaud.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        xr.DataArray: A mask raster generated from the polygons, representing the same geographical region as the source dataarray param
+                      0 where no polygon were found, and integers representing classes in order of occurence in the GeoDataFrame
+    """
+
+    # Find transformation
+    transform = np.array(data_array.spatial_ref.GeoTransform.split(" "), dtype=float)
+    transform = rasterio.transform.Affine.from_gdal(*transform)
+
+    # Find array shape
+    array_shape = data_array.shape[1:]
```
Comment (eschalkargans) [inline src/training_raster_clipper/implementation/nbertaud.py:96]: When reading this code, I know instictively that the shape is of len 3, and constituted of band, y, x, so that when you say `shape[1:]` what you really say is "shape of y and x dimensions, ignore the band"

You can be more explicit by saying for instance `data_array.isel(band=0).shape`

"isel with 0" means "take the first slice, whatever", then use the slice of it

so basically `1` became `band`

One of the main goal of xarray is to name the dimensions

File: src/training_raster_clipper/implementation/nbertaud.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        burnt_polygons (PolygonMask): Rasterized classified multipolygons
+
+    Returns:
+        _type_: A list of the RGB values contained in the data_array and their corresponding classes
+    """
+    # Flatten list in (x,y) dimension
+    band_array = data_array.stack(flatten=("y", "x"))
+    class_array = xr.DataArray(burnt_polygons.reshape(-1), dims="flatten")
+
+    # Fusion the arrays
```
Comment (eschalkargans) [inline src/training_raster_clipper/implementation/nbertaud.py:135]: This is not really a fusion but more a grouping
Now, these two variables share the same dimensions y and x. A Dataset is basically a dict of DataArrays sharing common dimensions

File: src/training_raster_clipper/implementation/nbertaud.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    """Extract RGB values covered by classified polygons
+
+    Args:
+        data_array (xr.DataArray): RGB raster
+        burnt_polygons (PolygonMask): Rasterized classified multipolygons
+
+    Returns:
+        _type_: A list of the RGB values contained in the data_array and their corresponding classes
+    """
+    # Flatten list in (x,y) dimension
```
Comment (eschalkargans) [inline src/training_raster_clipper/implementation/nbertaud.py:131]: (y,x), probably

So, xarray is nice BUT at the end it is good to recall what is the dimensions orders. From left to right, it is from least to most nested (like a C array)

File: src/training_raster_clipper/implementation/nbertaud.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    
+    # # Save the model
+    # joblib.dump(rf_clf, "path")
+    # rf_clf = joblib.load("path")
+
+    # Predict the classes of rasters
+    rasters_numpy = rasters.values
+    predictions = rf_clf.predict(rasters_numpy.reshape(4, -1).T)
+
+    predictions = predictions.reshape(rasters_numpy.shape[1:])
```
Comment (eschalkargans) [inline src/training_raster_clipper/implementation/nbertaud.py:185]: you can keep the shape before the prediction, with the `isel(band=0)`

**Review by eschalkargans — CHANGES_REQUESTED** [review]
Nice, I don't have that much to say, everything is formatted
My comments are more about "chore cleaning" and xarray

If you have some time, you can maybe try to generate outputs using the reference data contained in the tutorial `https://github.com/ARGANS/training-raster-clipper/tree/main/resources/solution/example_sentinel_files/SENTINEL.SAFE` (since you could not download the original because it was too old to be downloaded)

**Comment by Nicolas-Bertaud** [comment]
Can you review again please ? I have cleaned my files and use `isel(band=0)` instead of numpy shape with `[1:]`
